tools/outline_proxy_controller/outline_controller_server.cpp

```diff
@@ -23,74 +21,88 @@
 #include <unistd.h>
 #include <pwd.h>
 
-#include "outline_error.h"
+#include <boost/asio.hpp>
+#include <boost/lexical_cast.hpp>
+#include <boost/property_tree/json_parser.hpp>
+#include <boost/property_tree/ptree.hpp>
+
 #include "outline_controller_server.h"
+#include "outline_error.h"
 
 using namespace outline;
-using boost::asio::local::stream_protocol;
 
-static const char * const OUTLINE_GRP_NAME = "outlinevpn";
+//#region OutlineClientSession Implementation
 
-void session::start() {
-  auto self(shared_from_this());
-  boost::asio::spawn(executor_, [this, self](boost::asio::yield_context yield) {
-    try {
-      std::ostringstream response;
-      std::string clientCommand, buffer;
-      std::cout << "Client Connected" << std::endl;
-      for (;;) {
-        for (;;) {
-          boost::asio::async_read_until(socket_, boost::asio::dynamic_buffer(buffer, 1024), "}",
-                                        yield);
-          std::cout << buffer << std::endl;
-          clientCommand.append(buffer);
-          buffer.clear();
-          if (isValidJson(clientCommand)) {
-            std::cout << "Valid JSON" << std::endl;
-            break;
-          }
-        }
-        auto [err, result, action] = runClientCommand(clientCommand);
-        // TODO: replace the following code with a json library to handle special characters
-        response << "{\"statusCode\": " << err
-                 << ",\"returnValue\": \"" << result << "\""
-                 << ",\"action\": \"" << action << "\"}" << std::endl;
-        boost::asio::async_write(
-            socket_, boost::asio::buffer(response.str(), response.str().length()), yield);
-        std::cout << "Wrote back (" << response.str() << ") to unix socket" << std::endl;
-        clientCommand.clear();
-        response.str(std::string());
-      }
-    } catch (std::exception& e) {
-      socket_.close();
-    }
-  });
-}
+// Minimum length of JSON input from app
+static const int kJsonInputMinLength = 10;
 
-bool session::isValidJson(std::string str) {
-  std::stringstream ss;
-  boost::property_tree::ptree pt;
+// The buffer size used to communicate with Outline client
+static const int kChannelBufferSize = 1024;
 
-  if (str.length() >= JSON_INPUT_MIN_LENGTH) {
-    ss << str;
+/**
+ * @brief Checks the input string and returns true if it's a valid json.
+ */
+static bool IsValidJson(const std::string &input) {
+  if (input.length() >= kJsonInputMinLength) {
```

Comment:
Please move this to the caller. It doesn't have to do with whether it's a valid JSON, so it's confusing.

Reply:
I think this is for performance optimization, when we are receiving small portions of bytes, for example, a client request `{ "action": "test" }` might be devided into: `{ "act`, `ion": "`, `"test"`, ` }`. We won't waste any resources when we just received `{ "act` and `ion": "`.

Reply:
Done.